syz-verifier/monitoring_api.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+)
+
+// Monitor provides http based data for the syz-verifier monitoring.
+type Monitor struct {
+	externalStats *Stats
+}
+
+// MakeMonitor creates the Monitor instance.
+func MakeMonitor() *Monitor {
+	return &Monitor{}
+}
+
+// InitHTTPHandlers initializes the API routing.
+func (monitor *Monitor) InitHTTPHandlers() {
+	http.Handle("/api/stats.json", &statsHandler{
+		Src: monitor.externalStats,
+	})
+
+	http.HandleFunc("/", func(writer http.ResponseWriter, request *http.Request) {
+		writer.Write([]byte("<a href='api/stats.json'>stats_json</a>"))
+	})
+}
+
+// SetStatsTracking points Monitor to the Stats object to report.
+func (monitor *Monitor) SetStatsTracking(s *Stats) {
```

Comment:
Pass Stats as MakeMonitor argument.

Reply:
I expect per-VM monitoring later. Don't want MakeMonitor() to accept many parameters and dynamically create some complex structures to init Monitor().

Reply:
Why?
It's generally better to create objects in usable state with 1 call, rather than create objects in unusable state and then have additional methods to finish initialization.

Reply:
Partially agree.
Created object should be in valid state. "Useful" sounds arguable here.
After the creation, it is useful enough to set the tracking objects :)

My main concern here was the potential requirement to handle many parameters in "MakeMonitor()".
The more parameters you have - [the harder it is for the average developer to maintain or use this code](https://en.wikipedia.org/wiki/The_Magical_Number_Seven,_Plus_or_Minus_Two).

Reply:
Well, now there are 0 arguments, I proposed to add just 1. I can foresee there will be 2 - the second is an array of per-kernel objects. And in the end 7 arguments or 7 methods to call is not much different, it's still 7 objects to hold in short-term memory ;)